python/sunbeam/parser.py

```diff
@@ -1,118 +1,188 @@
-                "PARSE_UNKNOWN_KEYWORD"
-                "PARSE_RANDOM_TEXT"
-                "PARSE_RANDOM_SLASH"
-                "PARSE_MISSING_DIMS_KEYWORD"
-                "PARSE_EXTRA_DATA"
-                "PARSE_MISSING_INCLUDE"
-                "UNSUPPORTED_SCHEDULE_GEO_MODIFIER"
-                "UNSUPPORTED_COMPORD_TYPE"
-                "UNSUPPORTED_INITIAL_THPRES"
-                "UNSUPPORTED_TERMINATE_IF_BHP"
-                "INTERNAL_ERROR_UNINITIALIZED_THPRES"
-                "SUMMARY_UNKNOWN_WELL"
-                "SUMMARY_UNKNOWN_GROUP"
-            The avaiable recovery actions can be one of the following:
-                sunbeam.action.throw
-                sunbeam.action.warn
-                sunbeam.action.ignore
-
-    Examples:
-        Parses a deck from the string "RUNSPEC\\n\\nDIMENS\\n 2 2 1 /\\n"
-            deck = sunbeam.parse_deck("RUNSPEC\\n\\nDIMENS\\n 2 2 1 /\\n")
-        Parses a deck from the NORNE data set with recovery set to ignore
-        PARSE_RANDOM_SLASH error events.
-            deck = sunbeam.parse_deck('~/opm-data/norne/NORNE_ATW2013.DATA',
-                recovery=('PARSE_RANDOM_SLASH', sunbeam.action.ignore))
-
-    :rtype: sunbeam.libsunbeam.Deck
+
+    if not os.path.isfile(deck_file):
+        raise IOError("No such file: {}".format(deck_file))
+
+    context, parser = _init_parse(recovery, keywords)
+    return SunbeamState( lib.parse(deck_file, context, parser))
+
+
+
+def parse_string(deck_string, recovery=[], keywords=[]):
+    """Will parse a string and create SunbeamState object.
+
+    See function parse() for further details about return type and the recovery
+    and keyword arguments.
+
+    """
+    context, parser = _init_parse(recovery, keywords)
+    return SunbeamState(lib.parse_string(deck_string, context, parser))
+
+
+
+def deck(deck_file, keywords=[], recovery=[]):
+    """
+    Will parse a file and return a Deck object.
+
+    See function parse() for details about the keywords and recovery arguments.
     """
-    if keywords:
-        # this might be a single keyword dictionary, in which case we pack it
-        # into a list. If it's not a dict we assume it's an iterable and just
-        # carry on
-        if isinstance(keywords, dict):
-            keywords = [keywords]
-        keywords = list(map(json.dumps, keywords))
-    is_file = isfile(deck) # If the deck is a file, the deck is read from
-                           # that file. Otherwise it is assumed to be a
-                           # string representation of the the deck.
-    pc = _parse_context(recovery) if recovery else lib.ParseContext()
-    return lib.parse_deck(deck, keywords, is_file, pc)
+    if not os.path.isfile(deck_file):
+        raise IOError("No suc file: {}".format(deck_file))
+
+    context, parser = _init_parse(recovery, keywords)
+    return lib.create_deck(deck_file, context, parser)
+
+
+def deck_string(deck_string, recovery=[], keywords=[]):
```

Comment:
Why `deck_string` and not `parse_deck` as it was?

I also see you removed Jens' docstring, couldn't that be like it was?

Reply:
Of course it could have been like it was; I added quite a lot of documentation and changing/updating what was there already became part of that. 

My only semi technical argument to change it was that the list of available keys is changing, no reason to explicitly maintain the list manually in the docstring.

Reply:
> Why deck_string and not parse_deck as it was?

I do not really like any of these names - I guess it is a question of nouns versus verbs.

1. All of these functions *parse* a deck - that specifically apples to both the `parse()` and the `parse_deck() / deck()` functions.

2. The name `parse_deck()` seems to indicate that this function parses a deck - in contrast to `parse`which parses??? That is false - the distinction `parse_deck()` and `parse()` is that `parse_deck()` *returns* a deck - they operate on the same datastructure (i.e. a file).

In addition I have made one more change; the `parse()` function has been split in `parse()` and `parse_string()`. The previous implementation contained:
```python
if os.path.isfile(argument):
    parse_file()
else:
    parse_string()
```
that would mask a "/incorrect/path" argument; that will now raise `IOError`. The same for `deck() and `deck_string()`.

As I said I am not very happy with these names; but I think the current are better than what was there already.

Reply:
ping

Reply:
> [...] question of nouns versus verbs.

In my opinion, a function should preferably be a verb, and if it's difficult to figure out the "correct" name, we haven't really figured out how to design it.

What do you think about the following layout?
```python
deck = sunbeam.deck.parse(name)
state = sunbeam.parse(fname)
```

Reply:
```python
deck = sunbeam.deck.parse(name)
state = sunbeam.parse(fname)
```

I think that is good. We currently do not have any `sunbeam.deck` symbol, but we could of course introduce that.

I would prefer if  the main verb was  `load()` instead of `parse()` - it is e.g. `json.load()` and not `json.parse()`, similarly the csv functionality is typically in *readers*. 

We could actually get more inspiration from the json module - using  sunbeam.load() to load from a file and  sunbeam.loads() to load from a string.
```
# load from filename:
state = sunbeam.load("ECLIPSE.DATA")

#load from string:
state = sunbeam.loads( string_with_eclipse_input )
```

Reply:
I have no strong opinion on `parse` vs `load` vs `read`, but keep in mind that `load`, both for `json` and `yaml`, is _deserializing_ whereas reading/analyzing an eclipse deck is quite a lot heavier (see, e.g., _[Parsing a text file or data stream](https://stackoverflow.com/a/139246289)_).

However, if we can reach an agreement that way, I won't object to you changing the old use of `parse` with `load` (or even `read`).  (Though, it feels like it's on the side of the purpose of this PR.)

Reply:
OK - I now the api is:

```
import sunbeam

state = sunbeam.parse( )
state = sunbeam.parse_string( )
```

```
import sunbeam.deck

state = sunbeam.deck.parse()
state = sunbeam.deck.parse_string()
```